include/internal/quic_cc.h

```diff
@@ -14,146 +14,205 @@
+                        const OSSL_CC_LOSS_INFO *info);
 
     /*
-     * To be called when sent data was acked.
-     * |time_now| is current time in microseconds.
-     * |largest_pn_acked| is the largest packet number of the acked
-     * packets.
-     * |num_retransmittable_bytes| is the number of retransmittable
-     * packet bytes that were newly acked.
-     * Returns 1 if sending is unblocked (can_send returns 1), 0
-     * otherwise.
+     * To be called after a sequence of one or more on_data_lost() calls
+     * representing multiple packets in a single loss detection incident.
+     *
+     * Flags may be 0 or OSSL_CC_LOST_FLAG_PERSISTENT_CONGESTION.
      */
-    int (*on_data_acked)(OSSL_CC_DATA *ccdata,
-                         OSSL_TIME time_now,
-                         uint64_t last_pn_acked,
-                         uint64_t num_retransmittable_bytes);
+    int (*on_data_lost_finished)(OSSL_CC_DATA *ccdata, uint32_t flags);
 
     /*
-     * To be called when sent data is considered lost.
-     * |largest_pn_lost| is the largest packet number of the lost
-     * packets.
-     * |largest_pn_sent| is the largest packet number sent on this
-     * connection.
-     * |num_retransmittable_bytes| is the number of retransmittable
-     * packet bytes that are newly considered lost.
-     * |persistent_congestion| is 1 if the congestion is considered
-     * persistent (see RFC 9002 Section 7.6), 0 otherwise.
+     * For use when a PN space is invalidated or a packet must otherwise be
+     * 'undone' for congestion control purposes without acting as a loss signal.
+     * Only the size of the packet is needed.
      */
-    void (*on_data_lost)(OSSL_CC_DATA *ccdata,
-                         uint64_t largest_pn_lost,
-                         uint64_t largest_pn_sent,
-                         uint64_t num_retransmittable_bytes,
-                         int persistent_congestion);
+    int (*on_data_invalidated)(OSSL_CC_DATA *ccdata,
+                               uint64_t num_bytes);
 
     /*
-     * To be called when all lost data from the previous call to
-     * on_data_lost() was actually acknowledged.
-     * This reverts the size of the congestion window to the state
-     * before the on_data_lost() call.
-     * Returns 1 if sending is unblocked, 0 otherwise.
+     * Called from the ACKM when detecting an increased ECN-CE value in an ACK
+     * frame. This indicates congestion.
+     *
+     * Note that this differs from the RFC's conceptual segregation of the loss
+     * detection and congestion controller functions, as in our implementation
+     * the ACKM is responsible for detecting increases to ECN-CE and simply
+     * tells the congestion controller when ECN-triggered congestion has
+     * occurred. This allows a slightly more efficient implementation and
+     * narrower interface between the ACKM and CC.
      */
-    int (*on_spurious_congestion_event)(OSSL_CC_DATA *ccdata);
```

Comment:
Is the on_spurious_congestion_event() call something we do not want to provide?

Reply:
I'm unsure what this was referring to in MSQUIC as it's not a term which appears in RFC 9002. In any case we don't support it for now &mdash; so it will make more sense to add it if/when we have that functionality and have a better idea of what the API needs to look like. It would be a no-op if added for now anyway.

Reply:
It was about an event when we mark some packets as lost but then we later receive the lost packets. I.e. basically to undo last lost event. But yeah. This can be added later if we find that we need it.